### PR TITLE
fix: Table filter text loses focus when user is typing

### LINKED ISSUE #18040: [Bug]: Table Filter Text Loses Focus When User Is Typing
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

I have an appsmith app that has a table view with filtering enabled. If a user tried to filter one of the text fields (eg `Address contains XYZ`), the filter text input loses focus after a short delay after the user stops typing. This is causing UX issues for our users because if they type slowly, they have to click back into the text field to continue typing. It seems possible that the field is losing focus on debounce, tho we are not 100% sure. See the example below

appsmith-table-filters

### Steps To Reproduce

1. create a table widget with a text field
2. enable filtering
3. try to filter text column and stop typing for a fraction of a second.

### Public Sample App

_No response_

### Version

Appsmith Community v1.8.0

Front logo [Front conversations](https://app.frontapp.com/open/top_6jdzg)

### PR DESCRIPTION
**Pull request by souma-ghosh**
## Description

TL;DR
When user types in the input box for table filter, the input loses focus after a 500ms while typing. This is a regression from https://github.com/appsmithorg/appsmith/pull/16904 PR (bug - https://github.com/appsmithorg/appsmith/issues/12638) 

We were stringifying the whole filter object and using it in the key for each table filter. But when we changed the filter value the key would get updated and hence the filter re-rendered and the text input lost focus.
The key was previously set this way to address another bug where the value field of the filter would not update in the DOM although the filter data was correct.
The solution to address both these issues was to assign a unique key to the filters. Since the filter data is totally front-end generated we do not have any unique id to use here as the key. So we generate unique id for each filter now and use it as the key.

Fixes #18040 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

- Manual

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

**Review by sbalaji1192 — APPROVED** [review]

**Comment by kamakshibhat-appsmith** [comment]
@souma-ghosh The table crashes as soon as we remove dummy data from Data items .

**Review by sbalaji1192 — APPROVED** [review]

**Comment by souma-ghosh** [comment]
@laveena-en 
Page number should not reset when clearing a filter using Clear All btn or clearing a single filter with the cross btn in filter pane. Here is the loom explaining the behaviour using Clear All btn
https://www.loom.com/share/66019a500b4d4420b4c813823287897d

**Comment by laveena-en** [comment]
@souma-ghosh 
On adding an partial evaluation, the results being fetched are incorrect: https://www.loom.com/share/f051a226e2484f3db350c6e5cc882d28

Also, unable to navigate to the next page when a filter is applied:
https://www.loom.com/share/515c6e815a7640eaab504ae7ce0f16ac

**Comment by souma-ghosh** [comment]
> @souma-ghosh On adding an partial evaluation, the results being fetched are incorrect: https://www.loom.com/share/f051a226e2484f3db350c6e5cc882d28
> 
> Also, unable to navigate to the next page when a filter is applied: https://www.loom.com/share/515c6e815a7640eaab504ae7ce0f16ac

For the first one, the issue exists on release and prod. We should create an issue for this.
For the second one, if server side pagination is not integrated then this behaviour is correct as the number of results is less. That's why the the page count says 1 and the next button is disabled. Please correct me if I am wrong here.

**Comment by laveena-en** [comment]
Created a separate issue for the partial condition in filters: https://github.com/appsmithorg/appsmith/issues/20435

**Comment by laveena-en** [comment]
Tested and verified.
